contents/blog/building-a-new-feature-e2e.md

```diff
@@ -0,0 +1,81 @@
+
+## v0.5
+
+With a few insights supporting sampling, I then started to polish things a bit more. I asked our team to try out the feature, wrote some docs, and had a chat with our Head of Product Luke about what sampling should look like, how it should feel, etc. I also started to take Michael's UI feedback more seriously :D
+
+Having the docs early on was important, because it helped me scope out certain aspects of sampling more clearly, allowed our team to understand the feature better, and ensured it was ready to be turned on for customers in a second.
+
+But wasn't this only an internal feature at this point? 
+
+Indeed it was, but besides it being best practice to have some docs ready, we're generally very open with our customers about getting them on early tester programs - even when things are still quite rough - if we feel that we might be able to improve their experience a bit. This is also a fantastic way to get real feedback early on. 
+
+So soon enough, our support hero started tagging me in a few places where they felt sampling might help our customers out. Through this, we were both able to improve the experience for customers without having a feature fully ready, as well as we found new use cases for it that I hadn't thought of (e.g. sampling for experiments results, sampling on the toolbar).
+
+This is another reason why feature flags are so powerful: if a user requests access to a feature we're just testing out, it takes a few seconds to get it enabled for them.
+
+
+## v1 (Beta)
+
+Throughout the process, I made sure to demo the new things I was building at all-hands and by sharing demo videos on Slack, while constantly asking for feedback. The simple UI for the sampling selector changed four times, but ultimately we got to a state we were happy with. Things weren't perfect (the notice on dashboard panels still isn't great) but we we were happy putting this in front of users.
+
+I set up some instrumentation to track the feature's usage (share image of data we can share) and updated the feature flag to be rolled out to 25% of all organizations in PostHog. 
+
+As for analyzing usage and iterating, here's what my flow looks like now:
+
+- I keep key metrics in a dashboard and dig into them a few times a day
+- I watch session recordings of the feature being used to see if it's having the intended impact or causing any confusion (I've fixed a few visual regressions as a result of this)
+- I'm using an internal feature flag-based system we built to show a prompt calling for a user interview for the users that have tried sampling
+- The more confident I get that things aren't broken, I increase the rollout percentage a bit more on the flag and start from the top again
+
```

Comment:
For a conclusion, some questions you could answer:
1. What's good or bad (could be improved) about this process?
2. When are you "done" with a feature?

Reply:
Just to add to Ian's suggestions:

- I'd be interested to hear more about the key metrics you're using. Why did you choose them as a measure of success? This is probably something to add to this section.

For the conclusion, I'd be interested in:
1. Any feedback you've received from customers who have used it
2. Your future plans for this feature